Leetcode/aoc_2015_d2.py

```diff
@@ -0,0 +1,115 @@
+import unittest
+'''
+ ----- Part 1 -----
+Given dimensions of a rectangular square, get the total surface area
+with some additional slack which is determined by the area of the smallest
+side of all the dimensions given.
+e.x.
+2x3x4
+1x1x10
+expected output: 101
+surface area: 2*l*w + 2*w*h + 2*h*l
+dimensions 2x3x4
+l = 2, w = 3, h = 4
+2 *  6    + 2 *  12   + 2  * 8    = 52
+2 (2 * 3) + 2 (3 * 4) + 2 (4 * 2) = 52
+additional 6 ft of slack
+52 + 6 = 58
+
+The slack is found by getting the area of the smallest side.
+3 * 2 = 6
+
+
+The input gives us each package dimension in the format of
+LxWxH separated by a newline.
+To start off, we take in an input separated by lines and put
+it into a list.
+That way each line should look like:
+LxWxH
+LxWxH
+Then we'll further split the input by the xs so that only the numbers remain
+From there, it's just a matter of peforming the formula of
+2(l * w) + 2(w * h) + 2(l * h) + slack
+Putting the calculations for slack in a separate function makes it cleaner
+and easier to read.
+'''
+
+# Reads the input, splits it by lines, removes the newline,
+# removes the x's, leaving only numbers in a list.
+# converts those numbers from a string to an int.
+# so now packages is a 2D list.
+# e.x. 2x3x4 = [[2, 3, 4]]
+with open("aoc_2015_d2_input.txt") as file:
+	lines = file.readlines()
+	packages = [[int(x) for x in line.rstrip('\n').split('x')] for line in lines]
+
+
+# Sort the package list as the slack is the area of the smallest side.
+# Time complexity: Small portion of the input, doesn't add anything on its own.
+# however, due to the fact that it'll be run each time. Also sorts it. Should make it overall O(3n) -> O(n).
+def needed_slack(package: list) -> int:
+	package.sort()
+	return package[0] * package[1]
+
+# Time Complexity: Goes through the input at least once. O(n), runs the needed_slack. But that's a constant, so O(n) still.
+def needed_paper(packages: list) -> int:
+	total = 0
+	for p in packages:
+		surface_area = 2 * (p[0] * p[1]) + \
+		2 * (p[1] * p[2]) + 2 * (p[0] * p[2])
```

Comment:
Right, that makes sense. I noticed that it was actually fine on horizontal space, but I was coding this with half the screen and the other half with AoC on the other. Which is why I did that because at the time I didn't have much horizontal space.